File: frontend/src/types/workspace.ts
/**
 * 工作区类型定义
 */
import { ApiParam, DatabaseTable, DatabaseTableColumn, ApiInterfaceCard } from './modules';
import { UserSimple } from './user';

// 工作区基础类型
export interface Workspace {
  id: number;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  is_default: boolean;
  created_by: number;
  created_at: string;
  updated_at: string;
  user_count?: number;
}

// 工作区用户关系
export interface WorkspaceUser {
  user_id: number;
  workspace_id: number;
  access_level: string; // 'read', 'write', 'admin'
  username: string;
  email: string;
}

// 工作区数据库表
export interface WorkspaceTable {
  id: number;
  workspace_id: number;
  name: string;
  schema_name?: string;
  description?: string;
  columns_json: DatabaseTableColumn[];
  relationships_json?: any[];
  created_by: number;
  created_at: string;
  updated_at: string;
}

// 工作区数据库表详情
export interface WorkspaceTableDetail {
  id: number;
  workspace_id: number;
  name: string;
  schema_name?: string;
  description?: string;
  columns_json: DatabaseTableColumn[];
  relationships_json?: any[];
  columns: any[]; // 转换后的字段信息
  created_by: number;
  created_at: string;
  updated_at: string;
}

// 工作区接口
export interface WorkspaceInterface {
  id: number;
  workspace_id: number;
  path: string;
  method: string;
  description?: string;
  content_type?: string;
  request_params_json?: ApiParam[];
  response_params_json?: ApiParam[];
  request_example?: string;
  response_example?: string;
  created_at: string;
  updated_at: string;
  user_id: number;
}

// 工作区接口详情
export interface WorkspaceInterfaceDetail extends WorkspaceInterface {
  request_params: ApiParam[];
  response_params: ApiParam[];
}

// 创建工作区请求
export interface CreateWorkspaceRequest {
  name: string;
  description?: string;
  icon?: string;
  color?: string;
}

// 更新工作区请求
export interface UpdateWorkspaceRequest {
  name?: string;
  description?: string;
  icon?: string;
  color?: string;
}

// 添加用户到工作区请求
export interface AddUserToWorkspaceRequest {
  user_id: number;
  access_level: 'read' | 'write' | 'admin';
}

// 设置默认工作区请求
export interface SetDefaultWorkspaceRequest {
  workspace_id: number;
}

// 工作区列表响应
export type WorkspaceListResponse = Workspace[];

// 工作区用户列表响应
export type WorkspaceUserListResponse = WorkspaceUser[];

export interface CreateWorkspaceParams {
  name: string;
  description?: string;
  color?: string;
  is_default?: boolean;
}

export interface UpdateWorkspaceParams {
  name?: string;
  description?: string;
  color?: string;
}

export interface WorkspaceUserParams {
  user_id: number;
  /**
   * 用户在工作区的角色/访问级别
   * 注意：后端API使用access_level字段，但前端使用role字段显示
   * 
   * 角色映射关系：
   * - owner：完全控制权限，可以管理所有内容和设置
   * - admin：可以管理大多数内容和设置，但无法删除工作区
   * - member/write：可以查看和编辑内容，但无法更改关键设置
   * - guest/read：仅可查看权限，无法编辑内容
   */
  access_level: 'owner' | 'admin' | 'write' | 'read';
}

export interface BatchAddUsersToWorkspaceRequest {
  user_ids: number[];
  access_level: string;
}

export interface BatchRemoveUsersFromWorkspaceRequest {
  user_ids: number[];
}

export interface WorkspaceTableRead extends WorkspaceTable {
  creator?: UserSimple;
}

// 创建工作区表请求
export interface WorkspaceTableCreate {
  workspace_id: number;
  name: string;
  schema_name?: string;
  description?: string;
  columns_json: DatabaseTableColumn[];
  relationships_json?: any[];
}

// 更新工作区表请求
export interface WorkspaceTableUpdate {
  name?: string;
  schema_name?: string;
  description?: string;
  columns_json?: DatabaseTableColumn[];
  relationships_json?: any[];
}

// 创建工作区接口请求
export interface WorkspaceInterfaceCreate {
  workspace_id: number;
  path: string;
  method: string;
  description?: string;
  content_type?: string;
  request_params_json?: ApiParam[];
  response_params_json?: ApiParam[];
  request_example?: string;
  response_example?: string;
}

// 更新工作区接口请求
export interface WorkspaceInterfaceUpdate {
  path?: string;
  method?: string;
  description?: string;
  content_type?: string;
  request_params_json?: ApiParam[];
  response_params_json?: ApiParam[];
  request_example?: string;
  response_example?: string;
}

// 模块内容引用工作区表和接口的类型
export interface ModuleContentReferences {
  database_table_refs?: number[];
  api_interface_refs?: number[];
}

/**
 * 分页接口数据结构
 */
export interface PaginatedInterfaces {
  items: WorkspaceInterface[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 分页数据库表数据结构
 */
export interface PaginatedTables {
  items: WorkspaceTableRead[];
  total: number;
  page: number;
  page_size: number;
} 

File: backend/app/schemas/workspace.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime

from backend.app.schemas.module_content import DatabaseTableColumn, TableRelationship
from backend.app.schemas.user import UserSimpleRead


# 定义ApiParam模型，因为module_content中没有这个模型
class ApiParam(BaseModel):
    """API参数模型"""
    name: str
    type: str
    required: bool = False
    description: Optional[str] = None
    example: Optional[str] = None
    children: Optional[List['ApiParam']] = None


class WorkspaceBase(BaseModel):
    """工作区基础模型"""
    name: str
    description: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_default: Optional[bool] = False


class WorkspaceCreate(WorkspaceBase):
    """创建工作区请求模型"""
    pass


class WorkspaceUpdate(BaseModel):
    """更新工作区请求模型"""
    name: Optional[str] = None
    description: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_default: Optional[bool] = None
    default_prompt_template_id: Optional[int] = None


class WorkspaceInDB(WorkspaceBase):
    """数据库中的工作区模型"""
    id: int
    created_by: int
    created_at: datetime
    updated_at: datetime
    default_prompt_template_id: Optional[int] = None

    class Config:
        orm_mode = True


class Workspace(WorkspaceInDB):
    """工作区响应模型"""
    pass


# 工作区表模型
class WorkspaceTableBase(BaseModel):
    """工作区表基础模型"""
    name: str
    schema_name: Optional[str] = None
    description: Optional[str] = None
    columns_json: List[DatabaseTableColumn]
    relationships_json: Optional[List[TableRelationship]] = None


class WorkspaceTableCreate(WorkspaceTableBase):
    """创建工作区表请求模型"""
    workspace_id: int


class WorkspaceTableUpdate(BaseModel):
    """更新工作区表请求模型"""
    name: Optional[str] = None
    schema_name: Optional[str] = None
    description: Optional[str] = None
    columns_json: Optional[List[DatabaseTableColumn]] = None
    relationships_json: Optional[List[TableRelationship]] = None


class WorkspaceTableInDB(WorkspaceTableBase):
    """数据库中的工作区表模型"""
    id: int
    workspace_id: int
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class WorkspaceTable(WorkspaceTableInDB):
    """工作区表响应模型"""
    pass


class WorkspaceTableRead(WorkspaceTableBase):
    """用于读取的工作区表响应模型，包含创建者信息"""
    id: int
    created_at: datetime
    updated_at: datetime
    creator: Optional[UserSimpleRead] = None

    class Config:
        from_attributes = True


# 工作区接口模型
class WorkspaceInterfaceBase(BaseModel):
    """工作区接口基础模型"""
    path: str
    method: str
    description: Optional[str] = None
    content_type: Optional[str] = None
    request_params_json: Optional[List[ApiParam]] = None
    response_params_json: Optional[List[ApiParam]] = None
    request_example: Optional[str] = None  # 请求示例
    response_example: Optional[str] = None  # 响应示例


class WorkspaceInterfaceCreate(WorkspaceInterfaceBase):
    """创建工作区接口请求模型"""
    workspace_id: int


class WorkspaceInterfaceUpdate(BaseModel):
    """更新工作区接口请求模型"""
    path: Optional[str] = None
    method: Optional[str] = None
    description: Optional[str] = None
    content_type: Optional[str] = None
    request_params_json: Optional[List[ApiParam]] = None
    response_params_json: Optional[List[ApiParam]] = None
    request_example: Optional[str] = None  # 请求示例
    response_example: Optional[str] = None  # 响应示例


class WorkspaceInterfaceInDB(WorkspaceInterfaceBase):
    """数据库中的工作区接口模型"""
    id: int
    workspace_id: int
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class WorkspaceInterface(WorkspaceInterfaceInDB):
    """工作区接口响应模型"""
    pass


# 添加WorkspaceResponse别名以保持向后兼容
WorkspaceResponse = Workspace


class WorkspaceUserBase(BaseModel):
    """工作区用户基础模型"""
    user_id: int
    access_level: str = "read"


class WorkspaceAddUser(WorkspaceUserBase):
    """添加用户到工作区的请求模型"""
    pass


class WorkspaceBatchAddUsers(BaseModel):
    """批量添加用户到工作区的请求模型"""
    user_ids: List[int] = Field(..., description="要批量添加的用户ID列表")
    access_level: str = Field(default="read", description="统一设置的访问级别: read, write, admin")


class WorkspaceBatchRemoveUsers(BaseModel):
    """批量从工作区移除用户的请求模型"""
    user_ids: List[int] = Field(..., description="要批量移除的用户ID列表")


class WorkspaceUserResponse(WorkspaceUserBase):
    """工作区用户响应模型"""
    workspace_id: int
    username: str
    email: Optional[str] = None

    class Config:
        from_attributes = True


class WorkspaceWithUsers(WorkspaceResponse):
    """包含用户列表的工作区响应模型"""
    users: List[WorkspaceUserResponse] = []

    class Config:
        from_attributes = True


class UserDefaultWorkspace(BaseModel):
    """设置用户默认工作区的请求模型"""
    workspace_id: int


class WorkspaceUserRoleUpdate(BaseModel):
    """更新工作区用户角色的请求模型"""
    access_level: str = Field(..., description="用户访问级别: read, write, admin, owner等")